### PR TITLE
Masonry: Render tcm in position on first line

### PR DESCRIPTION
**Pull request by manvoid**
### Summary

Our current implementation on ssr for the two column modules position the item on the index defined on the array, f.e. if tcm index is 3 and it fits on the first row then it is rendered on that index. But this is not the case when doing a resizing, since instead of having a large batch we have the batches of 6 items and the tcm is sent to the first index of the batch always for the first line.

Since this two behaviors are different we can see that the position of the tcm changes greatly when doing resizing.

This PR adds a check to see if the heights are 0 and the tcm fits on the first line then it creates a batch starting on the tcm and skips the graph logic since it is not used, this way we can keep the position for the item on the first row.

### CODE REVIEW

File: packages/gestalt/src/Masonry/defaultTwoColumnModuleLayout.js [hunk 1]
```diff
@@ -302,8 +302,17 @@ const defaultTwoColumnModuleLayout = <T: { +[string]: mixed }>({
     if (hasTwoColumnItems) {
       // If the number of items to position is greater that the batch size
       // we identify the batch with the two column item and apply the graph only to those items
-      const splitIndex = calculateSplitIndex(itemsWithoutPositions);
-      const shouldBatchItems = itemsWithoutPositions.length > TWO_COL_ITEMS_MEASURE_BATCH_SIZE;
+      const twoColumnIndex = itemsWithoutPositions.indexOf(twoColumnItems[0]);
+
+      const skipGraph =
+        heights.every((height) => height === 0) && twoColumnIndex < heights.length - 1;
+      const shouldBatchItems =
```
Comment (liuyenwei) [inline packages/gestalt/src/Masonry/defaultTwoColumnModuleLayout.js:309]: if skipGraph is true do we still need to batch items? or is this just to keep the code simple since it doesn't really matter anyways

File: packages/gestalt/src/Masonry/defaultTwoColumnModuleLayout.js [hunk 2]
```diff
@@ -302,8 +302,17 @@ const defaultTwoColumnModuleLayout = <T: { +[string]: mixed }>({
     if (hasTwoColumnItems) {
       // If the number of items to position is greater that the batch size
       // we identify the batch with the two column item and apply the graph only to those items
-      const splitIndex = calculateSplitIndex(itemsWithoutPositions);
-      const shouldBatchItems = itemsWithoutPositions.length > TWO_COL_ITEMS_MEASURE_BATCH_SIZE;
+      const twoColumnIndex = itemsWithoutPositions.indexOf(twoColumnItems[0]);
+
+      const skipGraph =
+        heights.every((height) => height === 0) && twoColumnIndex < heights.length - 1;
```
Comment (liuyenwei) [inline packages/gestalt/src/Masonry/defaultTwoColumnModuleLayout.js:308]: more of an optimization but i don't think height specifically needs to be 0 here (e.g. we could say something like skip graph and just place where it is if the deviation of the heights in the array is less than the size of the gutter)

okay as a followup though. also, i don't think we need the check for twoColumnIndex < heights.lenght - 1?

File: packages/gestalt/src/Masonry/defaultTwoColumnModuleLayout.js [hunk 3]
```diff
@@ -402,15 +411,17 @@ const defaultTwoColumnModuleLayout = <T: { +[string]: mixed }>({
...
 
       // For each unpainted item, start generating possible layouts
-      oneColumnItems.forEach((item, i, arr) => {
-        addPossibleLayout({
-          item,
-          i,
-          arr,
-          heightsArr: paintedItemHeights,
-          prevNode: startNodeData,
+      if (!skipGraph) {
```
Comment (liuyenwei) [inline packages/gestalt/src/Masonry/defaultTwoColumnModuleLayout.js:414]: so just clarifying, we're still creating the graph with what's been painted so far but just not adding any layouts?

File: packages/gestalt/src/Masonry/defaultTwoColumnModuleLayout.js (hunk 2 above)
Comment (manvoid) [reply packages/gestalt/src/Masonry/defaultTwoColumnModuleLayout.js:308]: Yeah, I was thinking on how to apply the rule on a more general way but this is more accurate on what we want to solve at the moment. I can create a ticket to follow this as an algorithm improvement.

File: packages/gestalt/src/Masonry/defaultTwoColumnModuleLayout.js (hunk 1 above)
Comment (manvoid) [reply packages/gestalt/src/Masonry/defaultTwoColumnModuleLayout.js:309]: Exactly, its not really needed but we would execute the same steps as when there is a batch, I did it like this so we are not repeating code but that is another option on how to handle this.

File: packages/gestalt/src/Masonry/defaultTwoColumnModuleLayout.js (hunk 3 above)
Comment (manvoid) [reply packages/gestalt/src/Masonry/defaultTwoColumnModuleLayout.js:414]: Yeah, we create the graph and add the first node that is no layout. This is the one used as winning node and that puts the tcm right after painted items.

File: packages/gestalt/src/Masonry/defaultTwoColumnModuleLayout.js (hunk 2 above)
Comment (liuyenwei) [reply packages/gestalt/src/Masonry/defaultTwoColumnModuleLayout.js:308]: okay that sounds good - can we leave a comment here so we don't lose track of this? also, can you confirm we should be able to remove the `twoColumnIndex < heights.lenght - 1`

File: packages/gestalt/src/Masonry/defaultTwoColumnModuleLayout.js (hunk 1 above)
Comment (liuyenwei) [reply packages/gestalt/src/Masonry/defaultTwoColumnModuleLayout.js:309]: sg, can you leave a comment here around this?

**Review by liuyenwei — APPROVED** [review]

File: packages/gestalt/src/Masonry/defaultTwoColumnModuleLayout.js (hunk 2 above)
Comment (manvoid) [reply packages/gestalt/src/Masonry/defaultTwoColumnModuleLayout.js:308]: We can remove it, the side effect is that the item is going to be positioned as the next item always without calculating a better position, tbh that check doesn't solve that either so we could remove it. We would have to implement something to better position on that case.

File: packages/gestalt/src/Masonry/defaultTwoColumnModuleLayout.test.js [hunk 4]
```diff
@@ -388,7 +388,7 @@ describe('two column layout test cases', () => {
     const positionCache = new MeasurementStore<{ ... }, Position>();
     const heightsCache = new HeightsStore();
     const items = [
-      { 'name': 'Pin 0', 'height': 200, 'color': '#E230BA', 'columnSpan': 2 },
+      { 'name': 'Pin 0', 'height': 200, 'color': '#E230BA' },
```
Comment (liuyenwei) [inline packages/gestalt/src/Masonry/defaultTwoColumnModuleLayout.test.js:391]: why was this changed?

Comment (manvoid) [reply packages/gestalt/src/Masonry/defaultTwoColumnModuleLayout.test.js:391]: To test more positions, the same structure as the previous test, I'm testing when tcm is on index 0 and on index 3

File: packages/gestalt/src/Masonry/defaultTwoColumnModuleLayout.js [hunk 5]
```diff
@@ -40,6 +40,17 @@ function calculateTwoColumnModuleWidth(columnWidth: number, gutter: number): num
   return columnWidth * 2 + gutter;
 }
 
+function calculateSplitIndex(itemsWithoutPositionsLength: number, twoColumnIndex: number): ?number {
+  // If the items length is the same as the batch size we don't set a split index
+  if (itemsWithoutPositionsLength <= TWO_COL_ITEMS_MEASURE_BATCH_SIZE) {
+    return null;
```
Comment (liuyenwei) [inline packages/gestalt/src/Masonry/defaultTwoColumnModuleLayout.js:46]: nit: just return 0 here so return type can always be a number. no need to return null since 0 is also falsey

File: packages/gestalt/src/Masonry/defaultTwoColumnModuleLayout.js [hunk 6]
```diff
@@ -283,20 +294,21 @@ const defaultTwoColumnModuleLayout = <T: { +[string]: mixed }>({
...
-      const splitIndex =
-        twoColumnIndex + TWO_COL_ITEMS_MEASURE_BATCH_SIZE > itemsWithoutPositions.length
-          ? itemsWithoutPositions.length - TWO_COL_ITEMS_MEASURE_BATCH_SIZE
-          : twoColumnIndex;
-      const pre = shouldBatchItems ? itemsWithoutPositions.slice(0, splitIndex) : [];
-      const batchWithTwoColumnItems = shouldBatchItems
+      const splitIndex = skipGraph
+        ? twoColumnIndex
+        : calculateSplitIndex(itemsWithoutPositions.length, twoColumnIndex);
+      const pre = splitIndex ? itemsWithoutPositions.slice(0, splitIndex) : [];
```
Comment (liuyenwei) [inline packages/gestalt/src/Masonry/defaultTwoColumnModuleLayout.js:310]: if you change above to return 0, this can also just be `itemsWithoutPositions.slice(0, splitIndex)` since split(0,0) is the same as []

**Review by liuyenwei — APPROVED** [review]
